keps/sig-autoscaling/20181106-in-place-update-of-pod-resources.md

```diff
@@ -0,0 +1,376 @@
+in-place, without a need to restart the Pod or its Containers.
+
+The **core idea** behind the proposal is to make PodSpec mutable with regards to
+Resources, denoting **desired** resources.
+Additionally, PodStatus is extended to provide information about **actual**
+resource allocation.
+
+This document builds upon [proposal for live and in-place vertical scaling][] and
+[Vertical Resources Scaling in Kubernetes][].
+
+[proposal for live and in-place vertical scaling]: https://github.com/kubernetes/community/pull/1719
+[Vertical Resources Scaling in Kubernetes]: https://docs.google.com/document/d/18K-bl1EVsmJ04xeRq9o_vfY2GDgek6B6wmLjXw-kos4/edit?ts=5b96bf40
+
+## Motivation
+
+Resources allocated to a Pod's Container can require a change for various reasons:
+* load handled by the Pod has increased significantly and current resources are
+  not sufficient,
+* load has decreased significantly and allocated resources are unused and wasted,
+* resources have simply been set improperly.
+
+Currently, changing resource allocation requires the Pod to be recreated since
+the PodSpec's Container Resources is immutable.
+
+While many stateless workloads are designed to withstand such a disruption, some
+are more sensitive, especially when using low number of Pod replicas.
+
+Moreover, for stateful or batch workloads, a Pod restart is a serious
+disruption, resulting in lower availability or higher cost of running.
+
+Allowing Resources to be changed without recreating the Pod or restarting the
+Containers addresses this issue directly.
+
+### Goals
+
+* Primary: allow to change Pod resource requests & limits without restarting its
+  Containers.
+* Secondary: allow actors (users, VPA, StatefulSet, JobController) to decide
+  how to proceed if in-place resource resize is not possible.
+* Secondary: allow users to specify which Pods and Containers can be resized
+  without a restart.
+
+### Non-Goals
+
+The explicit non-goal of this KEP is to avoid controlling full life-cycle of a
+Pod which failed in-place resource resizing. This should be handled by actors
+which initiated the resizing.
+
+Other identified non-goals are:
+* allow to change Pod QoS class without a restart,
+* to change resources of Init Containers without a restart,
+* updating extended resources or any other resource types besides CPU, memory.
+
+## Proposal
+
+### API Changes
+
+PodSpec becomes mutable with regards to Container resources requests and limits.
+Additionally, PodSpec becomes a Pod subresource to allow fine-grained access control.
+
+PodSpec is extended with information about actually allocated Container resources.
+
+Thanks to the above:
+* PodSpec.Container.ResourceRequirements becomes purely a declaration, denoting
+  **desired** state of the Pod resources,
+* PodStatus.ContainerStatus.ResourcesAllocated (new object) shows the resources
+  held by the Pod and its Containers.
+
+In order to determine the state of a Pod resource resize, we add a new PodCondition
```

Comment:
We will need the FailedNodeCapacity condition so that we can quickly indicate to the initiating actor, such as VPA, that under current conditions it is not possible to resize. VPA may then choose to let it run, or evict the Pod to schedule to a new node that has enough resources, based on the policy configured by user.

Reply:
Some resizing behavior will take time (e.g., lowering the memory limit). How would this condition be set and what kind of latency is expected? It's possible that this is included in the KEP and I just missed the section. Let me know if that's the case.

Reply:
I've captured this in the Notes section of the KEP. In practice, initiating agent such as VPA should issue a resize lower request when the Pod memory use has measured lower than requested over a period of time as shown by metrics server and historical usage. Resetting the request value lower should happen when the resize hits HandlePodUpdates. If the application is really not holding more memory pages than VPA recommended limit, then UpdateContainerResources API should also return success. The concern is what do we do if it were to fail, and the suggestion is to set limits near actual usage to force reclaim. I suppose this could happen if the usage climbs after VPA makes the recommendation, and that recommendation is being applied.

Reply:
@yujuhong I looked into getting rid of this PodCondition, however without it, the only way for an initiating actor (VPA) to know whether a resize has failed (say due to node capacity) is to wait for a certain amount of time before possibly evicting the pod if allowed by policy. 

As it is currently designed, this is generated simply based on observation whether desired resource will fit the current node available resources, and if lost, it can be recovered. Having it helps initiating actor take alternative action much more rapidly, which better helps the overall cluster resource usage. I think we should keep it. Are there any specific corner cases you are worried about?

Regarding your concern with container restart if CRI fails, I agree we can just keep retrying the API instead. This is typical behavior of various K8s processes, so I've removed the restart-on-failure approach.

Let's discuss during tomorrow's sig-node meeting if we have a better way to determine currently allocated resources on Kubelet restart. I have another idea in mind - but don't like it too much.

Reply:
I am also strongly in favor removing the pod condition, as it has a number of issues:
1. The condition is not reconstructable from the state of the node.  This leads to many corner cases, as we have previously discussed.
1. It tries to fit events into a condition. I define an event as something that happens at a point in time, as opposed to conditions, which are a persistent, observable description of current state.
1. The kubelet's desired state for a pod should not be defined by previously-recorded conditions.

I disagree that enabling actors to determine cause of failure should be a goal.  I'm not sure why I would directly care about that as a user, and feels like an implementation detail rather than a goal.  You could make the case that users would like to be able to place time bounds on how long a scale up can take, but that can be accomplished with a timeout.

Reply:
Maybe I didn't explain clearly this morning... If we can find a reliable source-of-truth for PodStatus.ContainerStatus.ResourcesAllocated (either by persisting resource allocation locally on node or by moving it to PodSpec w/ subresource update), this PodCondition becomes fully reconstructable. This is set to Failed, InProgress, or Success depending on where kubelet is w.r.t handling requested resize. If kubelet were to restart, it looks at ResourceRequirements and ResourcesAllocated to regenerate the condition. The desired state for a pod w.r.t resize is stated only by the ResourceRequirements, nothing else.

The advantage of keeping this is that, by letting the initiating actor know resize is not possible now, or is being done, allows the initiating actor to take action such as evict the pod if allowed by policy. Alternately, the initiating actor could choose other pods it owns to evict (vs. kubelet or scheduler doing the preemption).

Without this, we effectively block out the desired capacity (due to max(Resources, ResourcesAllocated) while waiting for retry to eventually succeed or initiating actor chooses to evict the resize failed pod after a certain timeout. This leads to cluster resource wastage.

Reply:
Presumably, "moving it to the pod spec" actually means changing what the field is.  I.E. It stops being an observation of the current state of the container, as observed by the container runtime, and is now a statement of what the kubelet is reconciling towards.  I would think that even if we wanted to add this to the pod spec, we would still want to report the current state of the container according to the runtime.

Checkpointing should be avoided if at all possible, as discussed at sig-node.

I'm thinking out loud here, but it feels like the initiating controller is asking one question about the state of the pod, and a different question about the state of the node.  VPA wants to know "Has the pod's actual resources been reconciled to its desired resources?" to determine if it is Succeeded or not.  It can do this with just the resource requests and ResourcesAllocated.  Then, it also wants to know "Is the node oversubscribed?" to determine if the reason that it isn't succeeded is because there isn't enough room on the node.  What if we added a node condition, "Oversubscribed", which is true when the sum of pod resource requests exceeds node allocatable?

> Without this, we effectively block out the desired capacity (due to max(Resources, ResourcesAllocated) while waiting for retry to eventually succeed or initiating actor chooses to evict the resize failed pod after a certain timeout. This leads to cluster resource wastage.

This _could_ be seen as a feature, rather than a bug.  Say a pod is using more memory than requested, and is scaled up by VPA as a result, but there isn't enough room on the node to increase memory requests.  This prevents other pods from being scheduled to the node in the interim, which is good since the pod is using over its previous memory requests.  If a pod were to be scheduled there, it likely wouldn't get the resources it needed, since the original pod was using more than requested.

Reply:
I might have misunderstood, but I thought we discussed checkpointing in the context of how often we update ResourcesAllocated (i.e: do we post the update once when we finally converge to desired resource values, or do we post intermediate values as we drive towards desired resource values)

The 2nd choice I refer to above is to store the per-container requests & limits information (atomic-write once when fully converged) for each pod locally on the node. Should kubelet restart, it can discover this information (much like how it discovers and reconciles pods on the node), derive the PodCondition, and then update the PodStatus if necessary. If Spec Resources and Status ResourcesAllocated differ, it starts reconciling the resources towards desired state.

The 1st option of storing ResourcesAllocated in spec doesn't mean what kubelet is reconciling towards, rather what it has reconciled to. And we post once when we finish reconciling the resources to desired value.

I'm not sure which is better choice - or if there is a another alternative for source of truth about current resource allocations for the pod.

Providing the current pod resize status information via pod condition simply enables options for initiating actor, and can help drive towards better cluster resource usage in this case.

Reply:
Continuing our discussion from sig-node...

To restate my high-level concern with the pod condition, it is about consistency.  We are saying that it is fully reconstructible from other API objects, including the pod spec, pod status, node allocatable, and other pods assigned to the node.  In order for the condition to be correct, the pod condition should be updated whenever any of those things changes.  For example, if a different pod is evicted from the node, that pod's status should ideally be updated _as part of the operation of evicting the other pod_, as it now technically fits the nodes capacity.  Since implementing that type of consistency is infeasible, we are left with a condition that is potentially out-of-date and incorrect.  I generally prefer when actors can rely on the fundamental sources of truth rather than on second-hand observations of that truth.

For completeness, the node condition also suffers from this problem, and is only preferable to a pod condition because it is less complex to reason about and keep in-sync.

> The 1st option of storing ResourcesAllocated in spec doesn't mean what kubelet is reconciling towards, rather what it has reconciled to. And we post once when we finish reconciling the resources to desired value.

This doesn't seem right.  I had incorrectly assumed this field would be written after "admission" of the new resources by the kubelet, but before it was propagated to the container runtime to change actual state.   In that case it would be still be the state that the kubelet is trying to reconcile towards.  If it is the state it has already reconciled towards, it doesn't belong in the spec.

Before dawn cut us off at the meeting, we were discussing how to propagate the fact that the pod didn't fit on the node, and you were concerned that, for example, if the container runtime was slow to update the resources assigned to the pod, that the initiating actor couldn't differentiate between that and the pod not fitting.  If you wrote ResourcesAllocated after "admission", the initiating actor could differentiate between these two states.  If `Spec.ResourceRequirements != Spec.ResourcesAllocated`, then it is stuck waiting to be admitted.  However, if `Spec.ResourcesAllocated != Status.ResourcesAllocated`, then the kubelet is being slow to update cgroups.  With this, I don't think it is even necessary to have the node condition anymore.  VPA can just have a very short timeout (e.g. 3s) for waiting for `Spec.ResourceRequirements == Spec.ResourcesAllocated`, so that if it doesn't immediately fit, the pod gets rescheduled.  VPA could have a longer timeout for waiting for status to catch up to spec (e.g. 5 minutes) if it wants.

Reply:
@dashpole We can update the Pod conditions when node capacity becomes available as I mentioned [here](https://github.com/kubernetes/enhancements/pull/686#discussion_r301340572) so that it is not stale. It's actually quite simple to implement - when a Pod leaves (evicted or for other reasons) freeing up resources, walk the list of resize candidate Pods from HandlePodUpdates, and set their condition and queue it for resize if the capacity allows for it.

Node condition can be an useful measure for the initiating actor as well, although it is less granular than setting pod condition. Referring to the example from today's discussion, say Node has 4G mem, and has Pods P1 (2G) & P2 (1G). Now VPA sees both Pods reaching their allocated mem, and requests resize to say P1 - 3G, P2 - 2G. Now P1 can be resized (in progress), but not P2 (failed capacity). We set Node Oversubscribed condition, VPA sees that - but can't really tell if it should evict P1 or P2 after waiting for N seconds. It is better than having no condition at all, which forces very short timeouts.

However, as @dchen1107 suggested, Node condition or another mechanism that provides node usage information can be very useful thing to have for scheduler to make smarter decisions about future pods.

BTW, we don't have to implement Pod or Node condition in phase 1. We can give it more thought - things might get clearer once we get going.

Reply:
And yes from prior discussion with @derekwaynecarr writing ResourcesAllocated after 'admission' of resize (but before limits are updated by runtime) will be a problem when lowering resources - e.g scheduler might think node has room, whereas the truth is that node is still working towards reconciling lower. That's why we need to update ResourcesAllocated only after reconcile is complete.

Are you suggesting we have Spec.ResourcesAllocated as well as Status.ResourcesAllocated? (Instead of PodCondition). I was only looking to move ResourcesAllocated from Status to Spec to create a source of truth.

Reply:
> Are you suggesting we have Spec.ResourcesAllocated as well as Status.ResourcesAllocated? (Instead of PodCondition). I was only looking to move ResourcesAllocated from Status to Spec to create a source of truth.

That is what I was suggesting.  Regardless of whether we decide to add Spec.ResourcesAllocated, we should have Status.ResourcesAllocated, which is a direct observation of the actual state of cgroups.  We should have one "source of truth" for the kubelet's desired state which is entirely within the Pod.Spec, and another "source of truth" for the actual state which is entirely contained within the Pod.Status.  Our current problem stems from the fact that we don't actually have a source of truth for the desired state of the kubelet (i.e. `Spec.ResourcesAllocated`, just the desired state of the system as a whole (`Spec.ResourceRequirements`).  But that doesn't alleviate the need to report the actual state.

If we had both, we would solve both the scheduler issue, and any issues around kubelet restarts.

Reply:
> Our current problem stems from the fact that we don't actually have a source of truth for the desired state of the kubelet (i.e. `Spec.ResourcesAllocated`, just the desired state of the system as a whole (`Spec.ResourceRequirements`). 

Clarification: Do you mean we add Spec.ResourcesAllocated to be 'source of truth' for **current** state of kubelet?

Let me think through this.. at first glance - it would work. Only downside I see is that initiating-actor/VPA has to sift through the array of containers as opposed to just looking at a pod condition.

Reply:
> Do you mean we add Spec.ResourcesAllocated to be 'source of truth' for current state of kubelet?

If by **current** state you mean actual state, then no.  I should also probably refer to this as state _of the pod_ instead of state of the kubelet since controllers themselves don't have state.

Spec.ResourcesAllocated is still the **desired** state of the container, since it is used by the kubelet to determine what resource values _should be_ applied to containers, rather than an observation of what resource values _have been_ applied to containers.  It is true that setting Spec.ResourcesAllocated is a result of a computation by the kubelet, but I consider this step not to be a reflection of any actual state of the container, but a further specification of the Pod.

Reply:
This is confusing because Spec.Resources is the desired state (of the Pod). Let me rephrase to see if I understand..
Spec.Resources = Desired state (Request+Limit) of Pod by VPA/initiating-actor
Spec.ResourcesAllocated = Desired state of Pod by kubelet (Request+Limit written by kubelet via subresource after it accepts the resize during HandlePodUpdates)
Status.ResourcesAllocated = current state (Status.ResourcesAllocated.Requests always equals Spec.ResourcesAllocated.Requests, but Status.ResourcesAllocated.Limits may not equal Spec.ResourcesAllocated.Limits). 
Is this what you have in mind?

What do you think about the approach of locally storing (atomic file-write) a map of a pod's  container ResourceAllocations (after resize has been accepted, and after successful limit update for all containers)? This can be a source of truth of current ResourcesAllocated similar to how kubelet, upon restart, discovers currently running pods.

Reply:
Yeah, that looks mostly correct.  `Status.ResourcesAllocated.Requests always equals Spec.ResourcesAllocated.Requests, but Status.ResourcesAllocated.Limits may not equal Spec.ResourcesAllocated.Limits` isn't always guaranteed to be true.  We still need to wait for the container runtime to update cpu.shares, for example.  But that should happen very quickly.

> What do you think about the approach of locally storing (atomic file-write) a map of a pod's container ResourceAllocations (after resize has been accepted, and after successful limit update for all containers)? This can be a source of truth of current ResourcesAllocated similar to how kubelet, upon restart, discovers currently running pods.

This would work as well.  I think sig-node generally prefers to avoid checkpointing, but is another way to accomplish essentially the same thing as Spec.ResourcesAllocated.

Reply:
IIRC we last discussed checkpointing in the context of periodically updating ResourcesAllocated as we converge (i.e. If we are going X -> Y -> Z with desired state of Z, we update Y.
Yes, CRI update should happen quickly - but there were concerns in the case of lowering memory limit.

IMHO, if local atomic file-write works well, it seems to be the simpler option that doesn't add load to APIserver. Let's discuss this tomorrow and hopefully we can get an agreement.

Reply:
@dashpole Continuing last week's discussion (sorry i've been busy with another project)..

Going with Node-local ResourcesAllocated source-of-truth approach, Let’s say we have the following workflow:
* VPA sets new desired Pod Requests and Limits
* Kubelet, in HandlePodUpdate, checks new desired Pod Requests for fit.
    * If it does not fit, Kubelet sets FailedNodeCapacity condition (or let VPA SLA timeout)
    * If it fits, Kubelet ‘admits’ the request, and stores the Pods new desired []Ctr(Req+Llim) using atomic file-write, then sets InProgress condition.
        * The resources array is also stored during HandlePodAddition for new Pods upon successful admission.
    * It then calls CRI UpdateContainerResources to set the new desired resource limits. (This could take some time for lowering memory)
    * Upon completion of CRI UpdateContainerResources, it updates ContainerStatus.ResourcesAllocated (and sets Success condition)

I'll work on the 2+ pods case, below is how it looks for a single Pod.

**Kubelet restart analysis for Node with 4 CPUs (after all Pods received)**

Scenario 1: Single Pod
——————————- 
Case 1: P1(2) -> P1(3) 
KL ‘admits’ request, restarts before atomic file-write.
On restart, HandlePodAddition admits P1 at originally stored-ResourcesAllocated, writes resources, sets InProgress condition, continues to CRI update, and updates PodStatus.

Case 2: P1(2) -> P1(3)
KL ‘admits’ request, restarts after atomic file-write.
On restart, HandlePodAddition admits P1 at updated stored-ResourcesAllocated, writes resources, continues to CRI update, and updates PodStatus.

Case 3: P1(2) -> P1(5) 
KL rejects request, does not update local store, restarts before setting FailedNodeCapacity condition.
On restart, HandlePodAddition admits P1 at originally stored-ResourcesAllocated, sets FailedNodeCapacity condition.

Case 4: P1(2) -> P1(5)
KL rejects request, restarts after setting FailedNodeCapacity condition.
On restart, HandlePodAddition admits P1 at originally stored-ResourcesAllocated.

Case 5: P1(2) -> P1(3) during kubelet restart
Same as Case 1

Case 3: P1(2) -> P1(5)
Same as Case 3

Scenario 2: Two Pods
——————————-
... to be continued.

Reply:
Scenario 2: Two (or more) Pod Updates
——————————————————
Case 1: P1(1), P2(1). Update P1(1) -> P1(2)
KL ‘admits’ request, restarts before atomic file-write.
During restart P2(1)->P2(2)
On restart, HandlePodAddition admits P1 at originally stored-ResourcesAllocated, writes resources, sets InProgress condition, queues CRI update task, and updates PodStatus.
Then KL admits P2. P1 is updated before P2 based on lower ResourceVersion value.

Case 2: P1(1), P2(1). Update P1(1) -> P1(3)
KL ‘admits’ request, restarts before atomic file-write.
During restart P2(1)->P2(2)
On restart, HandlePodAddition admits P1 at originally stored-ResourcesAllocated, writes resources, sets InProgress condition, queues CRI update task, and updates PodStatus.
Then KL sets FailedNodeCapacity condition for P2. P1 is updated before P2 based on lower ResourceVersion value.

Case 3: P1(1), P2(1). Update P1(1) -> P1(2)
KL ‘admits’ request, restarts after atomic file-write.
During restart P2(1)->P2(2)
On restart, HandlePodAddition admits P1 at updated stored-ResourcesAllocated, writes resources, sets InProgress condition, queues CRI update task, and updates PodStatus.
Then KL admits P2. P1 is updated before P2 based on lower ResourceVersion value.

Case 4: P1(1), P2(1). Update P1(1) -> P1(3)
KL ‘admits’ request, restarts after atomic file-write.
During restart P2(1)->P2(2)
On restart, HandlePodAddition admits P1 at updated stored-ResourcesAllocated, writes resources, sets InProgress condition, queues CRI update task, and updates PodStatus.
Then KL sets FailedNodeCapacity condition for P2. P1 is updated before P2 based on lower ResourceVersion value.

Reply:
This looks like it works.  The only thing I am not sure about is being able to use resource version to order updates.  The kubelet has a goroutine for each pod to perform updates, so i'm not sure how you would ensure one is acted on before the other.  If we can order updates, I would suggest ordering them by priority rather than resource version.  But I don't think we need to guarantee an ordering to updates for this proposal to work.

We still need feedback on local checkpoint vs putting things in the pod spec

Reply:
As per our sig-node discussion this morning, I'm listing below the pros & cons that I can think of with saving accepted resize Locally vs PodSpec.

**Node-local storage of ResourcesAllocated via atomic file-write**:
- _Pros:_
  - Simpler (no change to PodSpec, avoids APIServer sub-resource code)
  - Faster (avoids PodSpec update RTT)
  - Extension of post-restart pod discovery - kubelet additionally discovers ResourceAllocation
- _Cons:_
  - Initiating actor cant see accepted ResourceAllocations
    - optional PodCondition helps provide this visibility

**PodSpec storage of ResourcesAllocated**:
- _Pros:_
  - Initiating actor can use PodSpec.ResourceAllocations to determine if resize was accepted
- _Cons:_
  - PodSpec becomes larger, needs more code for sub-resource and handling updates
  - Can be confusing
    - We would have pod.Spec.Container.Resources, pod.Spec.Container.ResourcesAllocated, and pod.Status.ContainerStatus.ResourcesAllocated

Reply:
> This looks like it works. The only thing I am not sure about is being able to use resource version to order updates. The kubelet has a goroutine for each pod to perform updates, so i'm not sure how you would ensure one is acted on before the other. If we can order updates, I would suggest ordering them by priority rather than resource version. But I don't think we need to guarantee an ordering to updates for this proposal to work.
> 
> We still need feedback on local checkpoint vs putting things in the pod spec

Upon restart, existing Pods are received as if they were new Pods and go through HandlePodAdditions. Assuming node-local storage, we discover existing Pods and their last granted ResourceAllocation values.

If the Pods received for APIserver show a resize request that  (Resources != local.ResourceAllocation), we queue it sorted by increasing ResourceVersion until all Pods have been received. Once all Pods have been received, we run them through HandlePodAdditions in the ResourceVersion sorted order. This should apply the resize in FCFS order that is resilient to Kubelet restarts.

Reply:
Long discussion on this issue, but short summary is that we didn't want Pod condition (now removed) and we are storing accepted resources in PodSpec. So resolving this issue.